src/librustc_typeck/collect.rs

```diff
@@ -1692,18 +1701,51 @@ fn fn_sig(tcx: TyCtxt<'_>, def_id: DefId) -> ty::PolyFnSig<'_> {
 
     match tcx.hir().get(hir_id) {
         TraitItem(hir::TraitItem {
-            node: TraitItemKind::Method(sig, _),
+            node: TraitItemKind::Method(MethodSig { header, decl }, TraitMethod::Provided(_)),
             ..
         })
         | ImplItem(hir::ImplItem {
-            node: ImplItemKind::Method(sig, _),
+            node: ImplItemKind::Method(MethodSig { header, decl }, _),
             ..
-        }) => AstConv::ty_of_fn(&icx, sig.header.unsafety, sig.header.abi, &sig.decl),
-
-        Item(hir::Item {
+        })
+        | Item(hir::Item {
             node: ItemKind::Fn(decl, header, _, _),
             ..
-        }) => AstConv::ty_of_fn(&icx, header.unsafety, header.abi, decl),
+        }) => {
+            get_infer_ret_ty(&decl.output).and_then(|ty| {
+                tcx.typeck_tables_of(def_id)
+                    .liberated_fn_sigs()
+                    .get(hir_id)
+                    .and_then(|fn_sig| Some((ty, *fn_sig)))
```

Comment:
I think you can just do `.liberated_fn_sigs()[hir_id]` - it should be impossible for it to be missing (@nikomatsakis would know more - but if tests pass, it should be fine).